master/build-rust-manifest.py

```diff
@@ -0,0 +1,489 @@
+            target_pkgs[target] = {
+                "available": True,
+                "url": url.replace(s3_addy, public_addy),
+                "hash": hash_from_s3_installer(url),
+            }
+        else:
+            print "error: " + name + " for " + target + " not available"
+            target_pkgs[target] = {
+                "available": False,
+                "url": "",
+                "hash": ""
+            }
+
+    return {
+        "version": version,
+        "target": target_pkgs
+    }
+
+# Find the URL of a package's installer tarball and return it or None if
+# the package doesn't exist on s3.
+#
+# NB: This method does not know how the installer is named: whether like
+# "rustc-nightly-$triple.tar.gz" or "rustc-$version-$triple.tar.gz". So
+# it will try both.
+def live_package_url(name, dist_dir, date, version, target):
+    # Cargo builds are always named 'nightly'
+    maybe_channel = channel
+    if name == "cargo":
+        maybe_channel = "nightly"
+
+    url1 = s3_addy + "/" + dist_dir + "/" + date + "/" + name + "-" + version + "-" + target + ".tar.gz"
+    url2 = s3_addy + "/" + dist_dir + "/" + date + "/" + name + "-" + maybe_channel + "-" + target + ".tar.gz"
+
+    print "checking " + url1
+    request = urllib2.Request(url1)
+    request.get_method = lambda: "HEAD"
+
+    try:
+        response = urllib2.urlopen(request)
+        if response.getcode() == 200:
+            return url1
+    except:
+        pass
+
+    print "checking " + url2
+    request = urllib2.Request(url2)
+    request.get_method = lambda: "HEAD"
+
+    try:
+        response = urllib2.urlopen(request)
+        if response.getcode() == 200:
+            return url2
+    except:
+        pass
+
+    return None
+
+# Finds the hash for an installer on s3 by downloading its sha256 file
+def hash_from_s3_installer(url):
+    hash_url = url + ".sha256"
+    print "retreiving hash from " + hash_url
+    response = urllib2.urlopen(hash_url)
+    if response.getcode() != 200:
+        raise Exception("expected hash not found at " + hash_url)
+
+    hash_file_str = response.read()
+    hash = hash_file_str[0:64]
+
+    return hash
+
+# Gets the url and hash of the 'rust' package for a target. These packages
+# have not been uploaded to s3, but instead are on the local disk.
+def url_and_hash_of_rust_package(target, rustc_short_version):
+    version = channel
+    if channel == "stable": version = rustc_short_version
+    file_name = "rust-" + version + "-" + target + ".tar.gz"
+    url = public_addy + "/dist/" + today + "/" + file_name
+    local_file = rust_package_dir + "/" + file_name
+    if not os.path.exists(local_file):
+        print "error: rust package missing: " + local_file
+        return None
+
+    hash = None
+    with open(local_file, 'rb') as f:
+        buf = f.read()
+        hash = hashlib.sha256(buf).hexdigest()
+
+    return {
+        "url": url,
+        "hash": hash,
+    }
+
+def write_manifest(manifest):
```

Comment:
`import pytoml` at the top, and making sure it's available in the environment where the buildmaster runs (`pip install pytoml` in the virtualenv. Switching to pytoml would be nice but a major refactor in order to build everything into an object for it to serialize